File: ios/Plugin/MicrophoneTypes.swift
//
//  MicrophoneTypes.swift
//  Plugin
//
//  Created by mozart alkhateeb on 30/05/2021.
//  Copyright © 2021 Max Lynch. All rights reserved.
//

struct AudioRecording {
    let path: String?
    let webPath: String?
    let duration: Int
    let format: String?
    let mimeType: String?

    func toDictionary() -> [String: Any] {
        var result: [String: Any] = [:]

        if path != nil {
            result["path"] = path
        }

        if webPath != nil {
            result["webPath"] = webPath
        }

        result["duration"] = duration

        if format != nil {
            result["format"] = format
        }

        if mimeType != nil {
            result["mimeType"] = mimeType
        }

        return result
    }
}

enum StatusMessageTypes: String {
    case microphonePermissionNotGranted = "microphone permission not granted"
    case cannotRecordOnThisPhone = "cannot record on this phone"
    case recordingFailed = "recording failed"
    case noRecordingInProgress = "no recording in progress"
    case failedToFetchRecording = "failed to fetch recording"
    case recordingInProgress = "recording in progress"
    case recordingPaused = "recording paused"
    case recordingResumed = "recording resumed"
    case microphoneIsBusy = "microphone is busy"
    case recordingStared = "recording stared"

    /// Correctly spelled alias of `recordingStared`.
    ///
    /// Swift forbids two cases sharing a raw value, so this is a static property
    /// rather than a case. The emitted string stays `"recording stared"`: it is
    /// part of the public contract and changing it would break existing callers.
    static let recordingStarted = StatusMessageTypes.recordingStared
}

enum MicrophonePermissionType: String, CaseIterable {
    case microphone
}

File: android/src/main/java/com/mozartec/capacitor/microphone/MicrophoneTypes.kt
package com.mozartec.capacitor.microphone

import com.getcapacitor.JSObject

enum class StatusMessageTypes(val value: String) {
    MicrophonePermissionNotGranted("microphone permission not granted"),
    CannotRecordOnThisPhone("cannot record on this phone"),
    RecordingFailed("recording failed"),
    NoRecordingInProgress("no recording in progress"),
    FailedToFetchRecording("failed to fetch recording"),
    RecordingInProgress("recording in progress"),
    RecordingPaused("recording paused"),
    RecordingResumed("recording resumed"),
    MicrophoneIsBusy("microphone is busy"),
    RecordingStared("recording stared");

    companion object {
        /**
         * Correctly spelled alias of [RecordingStared].
         *
         * Declared here rather than as a second entry so that iterating over
         * `values()` does not yield a duplicate. The emitted string stays
         * "recording stared": it is part of the public contract and changing it
         * would break existing callers.
         */
        @JvmField
        val RecordingStarted = RecordingStared
    }
}

class Recording(val path: String?, val webPath: String?, val duration: Int, val format: String?, val mimeType: String?) {
    fun toJSObject(): JSObject {
        var result = JSObject()

        if (path != null) {
            result.put("path", path)
        }

        if (webPath != null) {
            result.put("webPath", webPath)
        }

        result.put("duration", duration)

        if (format != null) {
            result.put("format", format)
        }

        if (mimeType != null) {
            result.put("mimeType", mimeType)
        }

        return result
    }
}